### ISSUE TITLE
Dasharo capsule updates

### ISSUE THREAD

**Issue by krystian-hebel** [Dasharo/dasharo-issues#423]
**The problem you're addressing (if any)**

Dasharo firmware currently is updated using flashrom. While it is a powerful tool that allows for a lot of customization, for end users it may be problematic. There is a lot of required options in the command line, what's worse they can differ between releases. This is both cumbersome and error prone.

Another issue is that flashrom requires access to the flash and its controller. This may be blocked by firmware, SMM and/or Secure Boot.

**Describe the solution you'd like**

The UEFI specification defines UEFI capsule services, `EFI_FIRMWARE_MANAGEMENT_PROTOCOL`, FMP capsule format and EFI System Resource Table (ESRT) to support system firmware and device firmware update. All of these contribute to capsule update mechanism, in which UEFI can update itself before any write protections are installed. Capsule can be passed to UEFI from OS (it's bootloader actually, since part of the process must be done before ExitBootServices() is called) or from storage device, which makes it possible to use Windows Update or fwupd to update firmware. Without going into details:

1. one or more update images are loaded into consecutive virtual memory,
2. virtual to physical memory mapping is passed to UEFI,
3. platform is rebooted,
4. UEFI finds and applies the firmware update.

That mechanism was designed with UEFI as main firmware in mind. Having coreboot running before it complicates few matters:

- After reboot, UEFI must protect the image from being overwritten. It does so by gathering all chunks of image(s) from where they were put by OS into one consecutive physical memory.
  - This memory is then excluded from PEI map of memory available for allocation, soon after MRC is done training RAM. With MRC called by coreboot instead of UEFI, either the gathering must be done by coreboot, or all small chunks of memory must be reserved. It is impossible to know where OS will put the capsules, so the latter option could theoretically end with no significantly big memory region to run ramstage or payload.
  - 64-bit OS running on top of 64-bit UEFI may load capsules above 4G. Currently coreboot isn't able to access that memory. In order to do so, at least a part of coreboot would have to run in 64-bit mode to gather all images and move them below 4G mark. Note that currently even 64-bit coreboot [maps only 4G of RAM](https://github.com/coreboot/coreboot/blob/master/src/cpu/x86/64bit/pt.S#L27).
  - Location and size of memory holding capsules after gathering must be reserved in coreboot and passed to UEFI payload in a HOB.
  - coreboot must somehow know that there are capsules to be parsed. EDK2 does so by [checking for existence of "CapsuleUpdateData" UEFI variable](https://github.com/tianocore/edk2/blob/edk2-stable202302/MdeModulePkg/Universal/CapsulePei/UefiCapsule.c#L867), ~~but coreboot is unaware of UEFI variables concept~~ **UPDATE:** it is now: https://review.coreboot.org/c/coreboot/+/52564. The same variable points to memory mapping, it can't be substituted with a single bit, so using only CMOS for this purpose is probably out of question.
- Access to flash is usually protected by SMM. SMI handlers are installed by coreboot and SMM memory is locked, which means that additional handler will have to be installed by coreboot.
  - This is similar in concept to existing SMMSTORE driver, except whole flash has to be writable. While reading is optional, validating that the flashing succeeded is a good practice. This may be done in SMM, without returning data back to the caller.
  - New handler should be secure, e.g. it shouldn't allow any software that can do `out` instruction write access to flash device. This most likely means that capsule signature must be validated by SMI handler instead of doing it in normal UEFI code. This in turn puts a requirement on SMM segment size. On the other hand, doing validation and flashing of whole image at once would result in few-seconds-long hang, which is bad for UX.
- On some platforms, vboot uses A/B update partitions. The idea behind them is that if an update of one region would result in non-booting or unstable platform, a second partition with previous version will be used as a fallback.
  - Ideally it should be done from OS after it is tested to be stable. To make it independent of user's input, firmware may set this flag just before it handles execution to OS loader, or when that loader calls ExitBootServices().
  - AFAICT there is currently no driver on UEFI side that would mark current boot as successful in vboot. It should be added to have a fully robust update mechanism.
  - Question as to when update the second partition is still open. Doing it together with marking boot as successful seems valid, but there won't be indication to users that the update is in progress.

**Where is the value to a user, and who might that user be?**

End users won't have to use flashrom for updating firmware. It will still be required for initial development.

Transition to fwupd and Windows Update in the future will be easier if capsule updates are supported.

**Describe alternatives you've considered**

Keep using flashrom?

**Additional context**

[This page](https://github.com/tianocore/tianocore.github.io/wiki/Capsule-Based-System-Firmware-Update) provides a nice overview of what has to be added on UEFI side. All of items linked there will have to be done, I won't repeat them here.

TODOs specific to coreboot:

- [x] implement access to >4G
- [x] implement gathering capsules into one region
- [x] reserve memory containing capsules and pass it to the payload
- [x] add SMI handler for flashing

TODOs specific to UEFI payload, other than those from link above:

- [x] parse version reported by coreboot to be put into ESRT
- [ ] support for vboot A/B scheme

Additional things I've noticed when trying to do PoC in QEMU:

- `EsrtDxe` mentioned in few places in wiki should be `EsrtFmpDxe`, both exist but the former doesn't seem to work.
- FDF and DSC must include `CapsulePei` and `VariablePei`, and for Ia32X64 also `Capsule64`. It isn't mentioned in [wiki](https://github.com/tianocore/tianocore.github.io/wiki/Capsule-Based-System-Firmware-Update-DSC-FDF). The module which sets boot mode must have `gPeiCapsulePpiGuid` in `[Depex]`. This probably won't be as important as in our case steps that are normally performed by `CapsulePei` will have to be done by coreboot instead.
- `PlatformBootManagerLib`: `ProcessCapsules()` has to be called twice, first time before `EndOfDxe` (i.e. while SMM is still unlocked) and second time after that (because some required drivers may be deferred until SMM is locked). This may impact UX, as for some platforms GOP is started after untrusted OptionROMs are executed.
- Special attention must be paid to GUIDs. Some of them (e.g. `PcdEdkiiSystemFirmwareFileGuid`) are specific to given platform, and jumping between wiki and code repositories may be error prone.
- `EdkiiSystemCapsuleLib` is unable to find `SystemFirmwareDescriptor` in PEI FV for OVMF. It may be caused by limited searching depth or by the fact that PEI+DXE FVs are compressed. I haven't dug into this, putting another copy of `SystemFirmwareDescriptor` in SEC allowed me to proceed (it must remain in PEI as well, ESRT needs it there).
  - coreboot version should also be reflected in it. Maybe it makes more sense to keep it outside the payload.
  - With A/B partitions there would be two copies of this information, choosing appropriate one in case of fallback to secondary slot requires careful design.

**Comment by miczyg1** [comment]
@krystian-hebel what should we do with this duplicate if we have this?: https://github.com/Dasharo/dasharo-issues/issues/124

**Comment by macpijan** [comment]
I know #124 was first, but this one is more complete, so I propose that we keep this one, and close the #124

**Comment by BeataZdunczyk** [comment]
> support for vboot A/B scheme

Instead of implementing support for the vboot A/B scheme, we decided to take a different approach.
We have focused on preserving user data during firmware updates, which is now fully implemented:
- coreboot PR: https://github.com/Dasharo/coreboot/pull/563
- edk2 PR: https://github.com/Dasharo/edk2/pull/167
- PR with test cases: https://github.com/Dasharo/open-source-firmware-validation/pull/501
- Test results: https://github.com/Dasharo/dasharo-issues/issues/809#issuecomment-2393991741


Closing this task as all planned Capsule Update integration has been done.
- [UEFI Capsule Update: Enable Capsule Updates in coreboot](https://github.com/Dasharo/dasharo-issues/milestone/31)
- [UEFI Capsule Update: Enable capsule updates in coreboot EDK II UEFI Payload](https://github.com/Dasharo/dasharo-issues/milestone/32)
- [UEFI Capsule Update: Test the solution on a hardware](https://github.com/Dasharo/dasharo-issues/milestone/33)
- [UEFI Capsule Update: Secure Firmware Signing](https://github.com/Dasharo/dasharo-issues/milestone/34)
- [UEFI Capsule Update: UEFI Capsule Update: Enhanced Boot Process with Preservation of User Data During Firmware Updates](https://github.com/Dasharo/dasharo-issues/milestone/35)
- [UEFI Capsule Update: Test and release UEFI Capsule Update for coreboot and EDK II](https://github.com/Dasharo/dasharo-issues/milestone/36)


We are now planning the Advanced UEFI Capsule Update for coreboot with EDK II.
